src/net_processing.cpp

```diff
@@ -599,14 +599,14 @@ class PeerManagerImpl final : public PeerManager
     std::atomic<int> m_best_height{-1};
 
     /** Next time to check for stale tip */
-    std::chrono::seconds m_stale_tip_check_time{0s};
+    std::chrono::seconds m_stale_tip_check_time GUARDED_BY(cs_main){0s};
```

Comment:
That it is *accessible* by other threads is why it needs to be guarded -- that's what ensures that it is only *accessed* by one thread at a time (or ever). Right now, if it were accessed by other threads who'd locked `cs_main`, it would be safe -- all this does is document that.

Reply:
Just because a variable is accessed under `cs_main` does not mean that it has to be protected by it. It would have to be if there was a relationship between `m_stale_tip_check_time` and some other variable that is already protected by `cs_main` and both have to be accessed together, but that is not the case.

This is the `m_stale_tip_check_time` access:

```cpp
if (now > m_stale_tip_check_time) {
    ...
    m_stale_tip_check_time = now + STALE_CHECK_INTERVAL;
}
```

If that would be called by multiple threads it would need a mutex protection. Not necessary `cs_main`. I think it is best to either leave it as is (because it is only accessed by a single thread) or introduce a new mutex for `m_stale_tip_check_time`.

Consider also a future work that tries to split `cs_main` to improve concurrency - then things have to be chopped off from `cs_main`. Needlessly adding `m_stale_tip_check_time` under it will create more work in the future.

Reply:
> Just because a variable is accessed under `cs_main` does not mean that it has to be protected by it.

That's true of everything protected by `cs_main` -- if it weren't there would be no hope of replacing cs_main by some other scheme. The issue isn't whether guarding the variable in some other way would also be safe, it's about documenting some way in which we can guarantee that the use is safe, and have the compiler maintain that guarantee for us.

Reply:
When I look at both `m_stale_tip_check_time` and `m_initial_sync_finished` and want to convince myself that there's no chance of one thread racing to access them while another is in the middle of modifying them, the way I convince myself of that is "oh, they're both used from precisely one function, and that function just locked cs_main".

It's certainly possible to look further, and say that `CheckForStaleTipAndEvictPeers` is only called via the scheduler thread; but that's more work, and it's not something that I can easily get the compiler to check for me. I *could* tell the compiler to check it for me with more work; but that involves adding thread-specific mutexes which reviewers here have objected to previously https://github.com/bitcoin/bitcoin/pull/21527#discussion_r631895780 and the whole point of breaking this PR out was to try to make some incremental progress in the meantime.